lib/dependencyutils.rb

```diff
@@ -0,0 +1,36 @@
+module DependencyUtils
+  def expand_dependencies (debug: false)
+    @checking = Array.new
+
+    def push_dependencies
+      if @pkg.is_binary?(@device[:architecture]) or \
+        !@pkg.in_upgrade and \
+        !@pkg.build_from_source
+    
+        # retrieve name of dependencies that doesn't contain :build tag
+        check_deps = @pkg.dependencies \
+          .reject {|k, v| v.include?(:build)} \
+          .map {|k, v| k}
+      else
+        # retrieve name of all dependencies
+        check_deps = @pkg.dependencies.map {|k, v| k}
+      end
+      # check all dependencies recursively
+      check_deps.each do |dep|
+        # build unique dependencies list
+        unless @dependencies.include?(dep) or \
+               dep == @pkgName or \
+               @checking.include?(dep) or \
+               installed?(dep)
+
+          search dep, true
+          @checking << @pkg.name
+          push_dependencies
+          @dependencies << dep
+        end
+      end
+    end
+    push_dependencies
+    remove_instance_variable(:@checking)
+  end
+end
```

Comment:
How does this ensure the dependencies are installed in the correct order?  For instance, is libpng installed before shared_mime_info or sommelier before its dependencies?

Reply:
Does `libpng` a dependency of them?

If current package depends on package A depends on pkgB depends on pkgC:
```
dependencies of current one:
pkgA
|- pkgB
   |- pkgC
```

The old algorithm will handle like this:
- Add pkgA to dependency array
- set current package to pkgA deps,each
- go back to step 1 but pkgA changed to pkgB
===snip===

Now it will handle like this:
- set current package to pkgA deps,each
- go back to step 1 but pkgA changed to pkgB
===snip===
- Add pkgA to dependency array

Reply:
shared_mime_info is a dependency of libpng so the install order should make shared_mime_info ahead of libpng.